collect_app/src/main/java/org/odk/collect/android/widgets/utilities/FormControllerWaitingForDataRegistry.java

```diff
@@ -0,0 +1,53 @@
+package org.odk.collect.android.widgets.utilities;
+
+import org.javarosa.core.model.FormIndex;
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.javarosawrapper.FormController;
+
+public class FormControllerWaitingForDataRegistry implements WaitingForDataRegistry {
```

Comment:
I don't think there's any reason for the `FormController` to be involved in this. I think that a single `WaitingForDataRegistry` object that keeps track of the state itself and is tied to the `FormEntryActivity` lifecycle would help reduce what the `FormController` deals with. Doesn't have to be done here but might be a nice next step if I'm not missing anything.